packages/rum-recorder/src/domain/segmentCollection.ts

```diff
@@ -0,0 +1,161 @@
+import { addEventListener, DOM_EVENT, EventEmitter, monitor } from '@datadog/browser-core'
+import { LifeCycle, LifeCycleEventType, ParentContexts, RumSession } from '@datadog/browser-rum-core'
+import { SEND_BEACON_BYTE_LENGTH_LIMIT } from '../transport/send'
+import { CreationReason, Record, SegmentContext, SegmentMeta } from '../types'
+import { DeflateSegmentWriter } from './deflateSegmentWriter'
+import { createDeflateWorker, DeflateWorker } from './deflateWorker'
+import { Segment } from './segment'
+
+export const MAX_SEGMENT_DURATION = 30_000
+
+// Segments are the main data structure for session replays. They contain context information used
+// for indexing or UI needs, and a list of records (RRWeb 'events', renamed to avoid confusing
+// namings). They are stored without any processing from the intake, and fetched one after the
+// other while a session is being replayed. Their encoding (deflate) are carefully crafted to allow
+// concatenating multiple segments together. Segments have a size overhead (meta), so our goal is to
+// build segments containing as much records as possible while complying with the various flush
+// strategies to guarantee a good replay quality.
+//
+// When the recording starts, a segment is initially created.  The segment is flushed (finalized and
+// sent) based on various events (non-exhaustive list):
+//
+// * the page visibility change or becomes to unload
+// * the segment duration reaches a limit
+// * the encoded segment size reaches a limit
+// * ...
+//
+// A segment cannot be created without its context.  If the RUM session ends and no session id is
+// available when creating a new segment, records will be ignored, until the session is renewed and
+// a new session id is available.
+//
+// Empty segments (segments with no record) aren't useful and should be ignored.
+//
+// To help investigate session replays issues, each segment is created with a "creation reason",
+// indicating why the session has been created.
+
+export function startSegmentCollection(
+  lifeCycle: LifeCycle,
+  applicationId: string,
+  session: RumSession,
+  parentContexts: ParentContexts,
+  send: (data: Uint8Array, meta: SegmentMeta) => void
+) {
+  const worker = createDeflateWorker()
+  return doStartSegmentCollection(
+    lifeCycle,
+    () => doGetSegmentContext(applicationId, session, parentContexts),
+    send,
+    worker
+  )
+}
+
+export function doStartSegmentCollection(
+  lifeCycle: LifeCycle,
+  getSegmentContext: () => SegmentContext | undefined,
+  send: (data: Uint8Array, meta: SegmentMeta) => void,
+  worker: DeflateWorker,
+  emitter: EventEmitter = window
+) {
+  let currentSegment: Segment | undefined
+  let currentSegmentExpirationTimeoutId: ReturnType<typeof setTimeout>
```

Comment:
Not systematically though: https://github.com/DataDog/browser-sdk/blob/master/packages/rum-core/src/domain/trackPageActivities.ts#L116

Reply:
maybe we could fix it then since it seems to be evaluated as `NodeJS.Timeout`

<img width="395" alt="Capture d’écran 2021-01-19 à 11 47 06" src="https://user-images.githubusercontent.com/1331991/105024195-13bb7a00-5a4c-11eb-9a02-d6b3df4e545d.png">

Reply:
I'll open a card for it, I'll apply your suggestion in the meantime